pkg/syncer/tunneler.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package syncer
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+
+	"github.com/kcp-dev/kcp/pkg/tunneler"
+)
+
+func startTunneler(ctx context.Context, upstream, downstream *rest.Config, syncTargetWorkspace logicalcluster.Name, syncTargetName string) error {
+	// syncer --> kcp
+	clientUpstream, err := rest.HTTPClientFor(upstream)
+	if err != nil {
+		return err
+	}
+
+	cfg := *downstream
+	// use http/1.1 to allow SPDY tunneling: pod exec, port-forward, ...
+	cfg.NextProtos = []string{"http/1.1"}
+	// syncer --> local apiserver
+	url, err := url.Parse(cfg.Host)
+	if err != nil {
+		return err
+	}
+
+	proxy := httputil.NewSingleHostReverseProxy(url)
+	if err != nil {
+		return err
+	}
+
+	clientDownstream, err := rest.HTTPClientFor(&cfg)
+	if err != nil {
+		return err
+	}
+	proxy.Transport = clientDownstream.Transport
+
+	u, err := url.Parse(upstream.Host)
+	if err != nil {
+		return err
+	}
+	// create the reverse connection
+	// virtual workspaces
+	dst := fmt.Sprintf("%s://%s/clusters/%s", u.Scheme, u.Host, syncTargetWorkspace)
```

Comment:
nit: `syncTargetWorkspace.Path()` gives you `/clusters/%s`

Reply:
fixed